### DISCUSSION TITLE
Python Pylint Import Error

### DISCUSSION THREAD

**Discussion by PancakePuncher** [community/community#46885] [Codespaces] [answered]
I have an issue where the VSCode in Codespaces is not correctly identifying imports that are installed.

The error is a Pylint error stating that a any of my imports cannot be imported, but this is inaccurate as the scripts run perfectly fine.

If I hover over the error there will be a description of the library/module that is accurate to indicate it was correctly imported, but a secondary error from Pylint saying it failed.

![Codespace-Import-Error-2](https://user-images.githubusercontent.com/3018272/217888310-bfae9622-75a8-4dbc-b1d9-d712c9267eae.png)

![Codespace-Import-Error](https://user-images.githubusercontent.com/3018272/217886455-0d3906fc-79d3-4edf-8d34-d189610b867d.png)

I have installed the VSCode Extensions and even tried with in both current and pre-release:

![Current-Exst](https://user-images.githubusercontent.com/3018272/217886775-80a7bfbd-766e-48f4-b362-6db2bd687f37.png)

I've attempted changing my interpreter to every possible option. I've used both Pyenv and Venv (Built-In). I've even switched to to the default 3.10.4 that was installed already on the container. I've insured the packages stating to be failing are installed on all Python versions.

![Inter-Settings](https://user-images.githubusercontent.com/3018272/217887328-ec5000f1-1eba-45fd-8dec-6b51690bc833.png)

It appears there is some kind of disconnect between what VSCode is using as the Interpreter and what Pylint is using.

VSCode Settings.json:

![image](https://user-images.githubusercontent.com/3018272/217888916-c5d12491-f96b-43db-8bb4-4bf1b5936b55.png)

I added the "python.terminal.activateEnvironment": false, settings based on the advice of other threads I found. These settings are my usual default settings on projects I work outside of Codespaces and don't have any issues.

I've even made sure this import works in the actual Python Interpreter.

![Interp](https://user-images.githubusercontent.com/3018272/217889431-94629f96-49d6-4452-8ef5-1a32df532388.png)

The script runs perfectly fine when I run it normally -- this seems to be strictly a visual error and not an actual error that keeps it from compiling and Interpreted.

I've tried many solutions to fix this, but nothing has worked. I've looked at others who reported this issue in the past and none of their solutions helped. I usually do everything I can to avoid submitting issues I know probably have been submitted before, but so far none of those threads have resolved this.

Any suggestions?

**Comment by mammaddrik** [comment]
## Quickstart
Install using pip:
`pip install uvicorn`
This will install uvicorn with minimal (pure Python) dependencies.
`pip install uvicorn[standard]`

## Answers
Add the directory name in front of your filename:
`uvicorn src.main:app`
or cd into that directory
`cd src`
`uvicorn main:app `

I'm not inside the same folder with my app instance, so if I try to run my app with uvicorn I'll get an error like yours.
The answer is so simple, add the folder name in front of your filename

I hope I was able to solve your problem.

**Reply by PancakePuncher** [reply]
This is not the issue. I can run uvicorn without issues using:

Uvicorn main:app --host 0.0.0.0 --port 8000

There is no issues in running the script.

The problem is that the pylint import errors will not go away no matter what I do.

The package is installed.

**Reply by mammaddrik** [reply]
Maybe this video will help you.
Youtube [Here](https://www.youtube.com/watch?v=BhNDn-FkKiU).

**Reply by PancakePuncher** [reply]
Unfortunately, I've already tried these steps. The error persist regardless of the Interpreter selected and Pylint is installed.

**Reply by mammaddrik** [reply]
Check this site too:
pylint [Here](https://pylint.readthedocs.io/en/latest/user_guide/messages/error/import-error.html).

**Reply by PancakePuncher** [reply]
I managed to solve it using ChatGBT -- I posted my fix in the global thread.

Thanks for your time!

**Comment by PancakePuncher** [comment, marked as answer]
This issue I have managed to fix it.

Using ChatGBT I was able to narrow down the issue to the VSCode setting for the path to Pylint being incorrectly pointed at a random Pylint stored inside the /usr directory.

After adjusting it to the Pylint in my Pyenv Environment settings and reloading the container. VSCode then re-installed Pylance (No idea why) and then upon a 2nd reload after that the issue is now resolved.

I do not understand why this was necessary and why adjusting the interpreter doesn't just override that setting. It has to be that Codespaces is forcefully pointing VSCode at the Pylint stored in /usr

For anyone who stumbles across this:

1. Hit "ctrl" + ","
2. Search for "python.linting.pylintPath"
3. Adjust the setting to wherever your Pylint is; usually "/path/to/python3.x/bin/pylint"
In my case it was "~/.pyenv/versions/3.10.9/bin/pylint"

**Reply by aniljatt** [reply]
Great!! So helpful :)

**Reply by iuritorres** [reply]
I love you, thank you, 2 days trying to resolve this lol.

**Reply by walawalagoose** [reply]
thx! I've already fixed this problem!

**Comment by jtcantin** [comment]
I had a similar issue. 

As I was moving through the steps given in [this answer](https://github.com/orgs/community/discussions/46885#discussioncomment-4924613) by [PancakePuncher](https://github.com/PancakePuncher), I found the setting "Pylint: Import Strategy". After installing pylint in my virtual environment, I set "Pylint: Import Strategy" to "fromEnvironment" and now the import resolution works correctly.
